src/Network/Wai/Log/Internal.hs

```diff
@@ -0,0 +1,39 @@
+{-# LANGUAGE RecordWildCards #-}
+module Network.Wai.Log.Internal where
+
+import Control.Monad (when)
+import Data.Aeson.Types (Value, object, emptyObject)
+import Data.Text (Text)
+import Data.Time.Clock (UTCTime, diffUTCTime, getCurrentTime)
+import Log (LogLevel)
+import Network.Wai (Middleware)
+
+import Network.Wai.Log.Options (Options(..), ResponseTime(..))
+
+-- | This type matches the one returned by 'getLoggerIO'
+type LoggerIO = UTCTime -> LogLevel -> Text -> Value -> IO ()
+
+-- | Create a logging 'Middleware' given a 'LoggerIO' logging function and 'Options'
+logRequestsWith :: LoggerIO -> Options -> Middleware
+logRequestsWith loggerIO Options{..} app req respond = do
+  logIO "Request received" . object . logRequest $ req
+  tStart <- getCurrentTime
+  app req $ \resp -> do
+    tEnd <- getCurrentTime
+    when logSendingResponse $
+         logIO_ "Sending response"
```

Comment:
That's mostly useless, because it has no reference to the request, and when you have multiple requests going at the same time, it's totally useless. Let's generate an ID for each request and add it to the 3 log events we have here, so that they're connected.

Reply:
> Let's generate an ID for each request

Agree, and I have been meaning to do that.
I first wanted to re-structure the code with options, before adding new features.
I not only want to generate an ID, but I want to be able to add that to the logging domain, so that it persists across the application.

Right now the (default) logging output has no changes compared to what was there previously, and that's very much on purpose.

We can make more changes after this step.

Reply:
Ok!